lib/elixir/lib/code/normalizer.ex

```diff
@@ -0,0 +1,246 @@
+  # The right hand side is an atom in the AST but it's not an atom literal, so
+  # it should not be wrapped
+  defp do_normalize({:., meta, [left, right]}, _parent_meta) do
+    {:., meta, [do_normalize(left, meta), right]}
+  end
+
+  # A list of left to right arrows is not considered as a list literal, so it's
+  # not wrapped
+  defp do_normalize([{:->, _, _} | _] = quoted, parent_meta) do
+    Enum.map(quoted, &do_normalize(&1, parent_meta))
+  end
+
+  # left -> right
+  defp do_normalize({:->, meta, [left, right]}, _parent_meta) do
+    left = Enum.map(left, &do_normalize(&1, meta))
+    right = do_normalize(right, meta)
+    {:->, meta, [left, right]}
+  end
+
+  # Maps
+  defp do_normalize({:%{}, meta, args}, _parent_meta) do
+    args = Enum.map(args, &do_normalize(&1, meta))
+
+    args =
+      case args do
+        # Unwrap the right hand side if we're in an update syntax
+        [{:|, pipe_meta, [left, {_, _, [right]}]}] ->
+          [{:|, pipe_meta, [left, right]}]
+
+        # Unwrap args so we have 2-tuples instead of blocks
+        [{_, _, args}] ->
+          args
+
+        args ->
+          args
+      end
+
+    {:%{}, meta, args}
+  end
+
+  # If a keyword list is an argument of a guard, we need to drop the block
+  # wrapping
+  defp do_normalize({:when, meta, args} = _quoted, _parent_meta) do
+    args =
+      Enum.map(args, fn
+        arg when is_list(arg) ->
+          {_, _, [arg]} = do_normalize(arg, meta)
+          arg
+
+        arg ->
+          do_normalize(arg, meta)
+      end)
+
+    {:when, meta, args}
+  end
+
+  # Calls
+  defp do_normalize({form, meta, args}, _parent_meta) when is_list(args) do
+    # Only normalize the form if it's a qualified call
+    form =
+      if is_atom(form) do
+        form
+      else
+        do_normalize(form, meta)
+      end
+
+    cond do
+      Keyword.has_key?(meta, :do) ->
+        {last_arg, leading_args} = List.pop_at(args, -1)
+
+        last_arg =
+          Enum.map(last_arg, fn {tag, block} ->
+            block = do_normalize(block, meta)
+
+            block =
+              case block do
+                {_, _, [[{:->, _, _} | _] = block]} -> block
+                block -> block
+              end
+
+            # Only wrap the tag if it isn't already wrapped
+            tag =
+              case tag do
+                {:__block__, _, _} -> tag
+                _ -> {:__block__, [line: meta[:line]], [tag]}
+              end
+
+            {tag, block}
+          end)
+
+        {_, _, [leading_args]} = do_normalize(leading_args, meta)
+
+        {form, meta, leading_args ++ [last_arg]}
+
+      true ->
+        args = Enum.map(args, &do_normalize(&1, meta))
+
+        {form, meta, args}
+    end
+  end
+
+  # Integers, floats, atoms
+  defp do_normalize(x, parent_meta) when is_literal(x) do
+    meta = [line: parent_meta[:line]]
+
+    meta =
+      if not is_atom(x) do
+        Keyword.put(meta, :token, Macro.to_string(x))
```

Comment:
Instead of defining it here, should we add a default for the formatter? Also, we probably want `inspect(x)` instead of Macro.to_string.

Reply:
Furthermore, :token is only needed for integer/float, so maybe we should split them into their own clause.